Analytical hessian for fe poly tensor

maieneuro opened the pull request:
Changed Hessian computations to analytical differentiation in FE_PolyTensor. This is the last remaining place where hessians are computed using finite differencing (as indicated by searching the code base for "compute_2nd", which now only appears in fe.h and fe.cc where it is declared and instantiated). 

I have changed the reference output of several test files. Unlike the previous analytical hessians patch, the differences are not just floating point errors. One salient issue is that the tests as they were before the change expected non-symmetric hessians, but the modified code produces symmetric ones. Is there a way to check correctness of the tests?

bangerth commented on include/deal.II/base/derivative_form.h:
```diff
@@ -97,7 +97,7 @@ class DerivativeForm
    * derivative is the Jacobian of F, then Tensor[i] = grad(F^i).
    */
-  operator Tensor<2,dim,Number>() const;
+  operator Tensor<order+1,dim,Number>() const;
```
You'll have to update the comment in front of the function.

bangerth commented on include/deal.II/base/derivative_form.h:
```diff
@@ -63,9 +63,9 @@ class DerivativeForm
+   * Constructor from a tensor.
    */
-  DerivativeForm (const Tensor<2,dim,Number> &);
+  DerivativeForm (const Tensor<order+1,dim,Number> &);
```
Nice improvement in itself. Can you add an entry to changes.h for just this?

In the future, feel free to isolate these changes to a single file and submit them as separate pull requests.

bangerth commented on source/fe/mapping_q1.cc:
```diff
@@ -1391,6 +1391,7 @@ MappingQ1<dim,spacedim>::transform (
                   }
 
               }
+      return;
```
Darn, the fact that we missed this `return` statement in the past means that nothing in the testsuite ever executes the block of code above it :-(

bangerth commented on tests/deal.II/fe_values_view_10.output:
```diff
@@ -6,25 +6,25 @@ DEAL::0 1.30 1.30 1.50
-DEAL::-1.43e+03 2.86e+03 2.65e+03 -2.63e+03
-DEAL::-1.43e+03 -2.60e+03 -2.12e+03 -1.44e+03
+DEAL::18.9 797. 797. -894.
+DEAL::18.9 -699. -699. -422.
```
Ouch, these are completely different values before and after :-(

bangerth commented:
I have only a few trivial comments on the patch itself.

Given that the results are so different, I would really like to add a few tests that check the correctness of the current code (it looks like the old code may not have worked correctly). The way to do this is to 

* Set up a test that simply computes the second derivatives for a Q1 element on a square. You can check that the result is correct by hand.
* Copy the test from step 1 into a separate test and repeat it for a Q2 element.
* Repeat these two steps for a mesh where you move two of the vertices of the square in a parallel direction, yielding an affine cell. Again, you can verify correctness by hand.
* Repeat by moving only a single vertex of the cell -- i.e., using a non-affine mapping.

If this works for the Q1 or Q2 element (which it should, of course, and I think you've already checked this), then repeat it all for elements that are derived from FE_PolyTensor. Ideally, since you have multiple code paths for the different transformation kinds, you'd do this for elements that test these different code paths (e.g., separate tests for RT and Nedelec elements; I don't know if there's an element derived from FE_PolyTensor that maps as either mapping_none, mapping_covariant, or mapping_contravariant, so that one can tests these code paths as well).

My style -- which I find easiest to debug -- is to write one test and then copy it with a minor modification, then copy that with another modification, etc, for each geometry and element I test. If you end up with a sequence of 20 little test programs (all of which will likely not have more than 20 or 30 lines of code), then that is totally fine with me. If you need some help with setting up test programs, take a look at
  http://dealii.org/developer/developers/testsuite.html
and feel free to ask for help!

kronbichler commented on include/deal.II/base/derivative_form.h:
```diff
@@ -180,17 +180,18 @@ DerivativeForm<order, dim, spacedim, Number>::DerivativeForm  ()
 
 
 
+
```
please delete this additional (fourth) empty line

maieneuro replied:
Done.

maieneuro commented:
I've added a test for the hessians using the divergence theorem, and sadly it fails in FE_RaviartThomas and FE_Nedelec. I think I know what the error is, and I think it's also there in the computation of the shape_gradients. If the values are computed using a mapping of the reference shape functions, shouldn't the gradients contain a term corresponding to the gradient of that mapping?

bangerth commented:
Clever test -- I like the technique a lot since you don't actually need to know the correct values of the shape functions or their hessians!

As for the error: test this out by cloning your test and applying it to a parallelogram there (but keep the old test -- as I said, the more tests the better!). That's a linear mapping and everything there should be ok. I suspect that you're probably right with your suspicion, but I don't know enough about how the RT element is defined to tell for sure.

What happens if you do the same for the Nedelec element. Does that work?

maieneuro commented:
Nedelec element also fails the test. Both the RT and Nedelec elements pass the test on a simple geoemtry (I tried hyper_cube) but both fail with hyper_ball. I've been trying different tests by modifying the same file but I should really make a copy of a test before modifying it. In any case once the code is modified so the test passes the output will be different, so I can remake all the tests at that stage.

agrayver commented:
@maieneuro I might be wrong, but with with hyper_cube you're getting a triangulation which has non-standard faces (see glossary on this). Those are definitely not supported by FE_Nedelec and likely by RT too. You can try hyper_shell instead, although there may also be some problems if faces at a boundary are not rectangles. You are probably better off avoiding spherical domains for the time being and follow Wolfgang's suggestion. 

P.S. Great that you're looking into this! I can assist with testing FE_Nedelec, but know little about RT unfortunately.

bangerth commented:
That's a valid point about the hyperball. I would try to use meshes that consist of only one cell and are purposefully deformed.

As for writing tests: yes, copy the test everytime you make a modification. When you fix the code, all you have to do is update the test output (and verify that it is now correct) but you won't have to touch the .cc file any more.

maieneuro commented:
My suspicion was right, there as a missing term in the derivatives. I've added the missing term and added tests using a version of the divergence theorem: and they pass. I still need to add the corresponding term to the hessians. I might be wrong, but by looking at the code, the gradients and hessians only depend on the reference shape gradients and hessians and those of the mapping, which don't depend on face orientation. This is also supported by the fact that both the nedelec and rt elements now pass the divergence theorem gradient tests on 2D and 3D hyper_ball meshes.

@agrayver more tests for gradients and hessians on FE_Nedelec would be nice.

bangerth commented:
@maieneuro -- good work! Fascinating (and frustrating/scary) that we had this bug for more than a decade and nobody noticed :-(

I believe that the hyper_ball in 3d has all correct edges but not faces. This would support the theory that the Nedelec element passes but the RT element doesn't. I know from discussions with Ross Kynch that we're missing a factor of -1 or similar for those cases somewhere, but we have not been able to identify the place where this might be.

Do you think you can break the changes to the RT and Nedelec elements out into a separate pull request (including an entry to changes.h)? It's often helpful to have separate things committed separately to make later searches through the repository to find the place where something changed simpler.

maieneuro commented:
@bangerth that makes sense, thanks for clearing that out. Perhaps in the future using the divergence theorem tests on hyper_ball for rt elements could help with verifying corrections to the missing factor?

I'll generate separate pull requests for the changes in DerivativeForm and the for the gradient computations along with tests. As for this PR, I'll keep it for its declared purpose: the analytical hessians. The missing terms in the analytical hessians involve the jacobian_hessian, which I'll implement next as it's anyway a step towards my main goal: the shape_hessian_grads.

rosskynch commented:
Fantastic work @maieneuro!

For the Nedelec element, I'd be interested to see if it fails on a cylinder? I'm a little surprised that it passes on the hyper_ball as I was fairly sure that the sign flip was a problem on non-standard faces and that edges were handled correctly.

I'm also wondering if the bug fixed in MappingQ1 may fix issues myself and @agrayver have experienced ( see #338) with MappingQ, as they call the Q1 transform.

maieneuro commented:
@rosskynch  I doubt the bug in MappingQ1 is the cause of that issue since the bug is in a new function whose interface didn't exist and, as @bangerth pointed out, isn't being called by anything at the moment (except in my fork for this PR and for the forthcoming hessian gradient computations patch).

rosskynch commented:
Ah apologies - I glanced at the functions quickly!

bangerth commented:
On 08/19/2015 09:00 AM, Maien Hamed wrote:
> @bangerth <https://github.com/bangerth> that makes sense, thanks for
> clearing that out. Perhaps in the future using the divergence theorem
> tests on hyper_ball for rt elements could help with verifying
> corrections to the missing factor?

Yes, it's exactly this kind of test that's very valuable because one 
doesn't actually have to know the exact solution. Having to know the 
exact solution is often a big hurdle to writing tests.

bangerth commented:
With #1394 merged, can you now rebase this patch here to master?

rosskynch commented:
One suggestion for FE_Nedelec testing:

The divergence theorem may not flag up a problem with non-standard meshes for this element - I've just had a short discussion with a colleague and we couldn't reach a conclusion so apologies if I'm mistaken here. One thought is to try the equivalent identity with curl (the name of the identity escapes me):

\int_{d\Omega} n \times A dS = \int_{\Omega} \nabla \times A d\Omega

maieneuro commented:
@rosskynch thanks, that would be a nice test to add. On the other hand I think the gradient theorem-based tests that are used here are actually a general case because they test the components of the gradient one at a time. The rule would be 
\int_{\Omega} \frac{\partial A}{\partial x_i} d\Omega = \int_{dS} A \otimes  e_i \cdot dS

I think the curl test can be recovered from a combination of these components?

bangerth commented:
I ran the tests on one of my machines, and these are the ones that currently fail for me:
```
        3144 - deal.II/fe_values_view_10.debug (Failed)
        3145 - deal.II/fe_values_view_10.release (Failed)
        3206 - deal.II/rt_covariant.debug (Failed)
        3207 - deal.II/rt_covariant.release (Failed)
        3210 - deal.II/fe_values_view_04.debug (Failed)
        3211 - deal.II/fe_values_view_04.release (Failed)
        3398 - deal.II/fe_values_view_10_single_02.debug (Failed)
        3399 - deal.II/fe_values_view_10_single_02.release (Failed)
        3400 - deal.II/fe_values_view_16.debug (Failed)
        3401 - deal.II/fe_values_view_16.release (Failed)
        3443 - deal.II/fe_values_view_03.debug (Failed)
        3444 - deal.II/fe_values_view_03.release (Failed)
        3485 - deal.II/fe_values_view_10_single_01.debug (Failed)
        3486 - deal.II/fe_values_view_10_single_01.release (Failed)
        3567 - deal.II/fe_values_view_10_single_03.debug (Failed)
        3568 - deal.II/fe_values_view_10_single_03.release (Failed)
        3735 - deal.II/rt_hessian.debug (Failed)
        3736 - deal.II/rt_hessian.release (Failed)
        4185 - fe/fe_nedelec_hessian_divergence_theorem.debug (Failed)
        4186 - fe/fe_nedelec_hessian_divergence_theorem.release (Failed)
```
Do you see these as well?

maieneuro commented:
yes, I see them. The terms that arise from the newly introduced term to the gradients still need to be added here. But those terms involve the Jacobian hessian, which I'm working on now and will submit as a separate PR. The fe_values_views_* tests will likely change, and some more divergence tests need to be added. But I'll leave that till I've added the missing terms. Should I close this PR until it's ready?

bangerth commented:
Your choice. It's ok to close and later re-open once you have all of the pieces in place, or you can keep it open and keep the `[WIP]` marker in the subject to indicate that it's not ready for merging yet.

maieneuro commented:
OK I think it's better to leave it open with the [WIP] marker in case any coments come up..

bangerth commented:
@maieneuro -- let us know when this is ready for re-review.

maieneuro commented:
This is dependent on the changes in PR #1441 (the missing terms need the 'soon to be renamed' jacobian_pushed_forward_hessian). So it makes most sense to add them after that PR is in its final revision.

maieneuro commented:
This is now ready for re-review

maieneuro commented on source/fe/mapping_q1.cc:
```diff
@@ -1116,8 +1116,8 @@ namespace internal
-                                  for (unsigned int mr=0; mr<mr_start; ++mr)
+                                  const unsigned int mr0 = (jr+lr == 0)? 1:0;
+                                  for (unsigned int mr=mr0; mr<dim; ++mr)
                                     jacobian_pushed_forward_2nd_derivatives[point][i][j][l][m]
```
What a silly bug, pasted in the wrong place! sorry about that.

bangerth commented on include/deal.II/fe/fe_abf.h:
```diff
@@ -173,28 +173,6 @@ class FE_ABF : public FE_PolyTensor<PolynomialsABF<dim>, dim>
-   *
-   * For the present kind of finite element, this is exactly the case.
-   */
-  virtual UpdateFlags update_once (const UpdateFlags flags) const;
```
Do you no longer need these functions? If not, why not?

bangerth replied:
Ah, because `FE_PolyTensor` has these functions that already provide the same service?

maieneuro replied:
Yes, exactly.. FE_PolyTensor is responsible for doing the computations and it knows the flags required.

maieneuro commented:
It seems like good practice to put the equations somewhere for reference, so here's a document https://drive.google.com/file/d/0B-7mnFLpdvqxR2k1cEUxMWFIZUk/view?usp=sharing. Is there a way to put the raw TeX here and disabling markdown?

bangerth commented:
Great idea with the formulas. I think the best approach would be to add them to the documentation of the `Mapping` class, to define the terms you are using elsewhere (such as "Jacobian" or "Hessian" of the mapping). Some of the formulas are already there:
https://www.dealii.org/developer/doxygen/deal.II/classMapping.html
Feel free to expand (and replace) the section on "Mathematics of the mapping" into a section that introduces terminology. You can include the latex formulas verbatim.

bangerth commented:
I will admit that I did not check every expression in the test, but I do like the fact that you wrote tests that are based on a mathematical equality.

Let me merge this patch. If you'd like to add the formulas somewhere (which I think would be great), then please do so in a follow-up pull request.

bangerth commented:
...did not check every expression in the test... -> did not check every expression in the patch...

maieneuro commented:
Thanks. I'll add the formulas in a future PR.

agrayver commented:
@maieneuro do I understand it right that with this patch, Nedelec elements should compute gradients correctly on non-rectangular cells? See "Restriction on transformations" on FE_Nedelec doc page.

If this is the case, then I do not observe the correct behavior even with your patch...

maieneuro commented:
@agrayver Yes, I believe that the missing terms described in "Restriction on transformations" are the ones that have been added in this patch. There is still, however, the issue of the sign flip discussed above. Could you please elaborate on how you're not observing the right behaviour? Is it perhaps something that can be used as a test case?

agrayver commented:
All faces in the mesh have standard orientations -- this is not the reason. I will construct a small example and create a separate issue next days.

maieneuro commented:
Good, thank you.
